File: docmanage_mobile/lib/api/category_api.dart
import 'http_service.dart';

class CategoryApi {

  static Future<Map<String, dynamic>> getCategories() async {
    try {
      final response = await HttpService.request(
        '/categories',
        method: 'GET',
      );

      return response;

    } catch (e) {

      return {
        "success": false,
        "message": e.toString()
      };

    }


  }

  static Future<Map<String, dynamic>> getCategory(int categoryId) async {
    try {

      final response = await HttpService.request(
        '/categories/$categoryId',
        method: 'GET',
      );

      return response;

    } catch (e) {

      return {
        "success": false,
        "message": e.toString()
      };

    }


  }

  static Future<Map<String, dynamic>> createCategory({
    required String name,
    String description = '',
  }) async {
    try {

      final response = await HttpService.request(
        '/categories',
        method: 'POST',
        body: {
          "name": name,
          "description": description
        },
      );

      return response;

    } catch (e) {

      return {
        "success": false,
        "message": e.toString()
      };

    }

  }

  static Future<Map<String, dynamic>> updateCategory({
    required int categoryId,
    required String name,
    String description = '',
  }) async {
    try {

      final response = await HttpService.request(
        '/categories/$categoryId',
        method: 'PUT',
        body: {
          "name": name,
          "description": description
        },
      );

      return response;

    } catch (e) {

      return {
        "success": false,
        "message": e.toString()
      };

    }


  }

  static Future<Map<String, dynamic>> deleteCategory(int categoryId) async {
    try {

      final response = await HttpService.request(
        '/categories/$categoryId',
        method: 'DELETE',
      );

      return response;

    } catch (e) {

      return {
        "success": false,
        "message": e.toString()
      };

    }


  }


}

File: frontend/src/api/categoryAPI.js
import { http } from './http.js';

export async function getCategories(){
    try {
        const data = await http('/categories');
        return data;
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function getCategory(id){
    try {
        const data = await http (`/categories/${id}`, 
        {
            method: "Get",
        });
        return data;
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function createCategory(payload){
    const { name, description = '' } = typeof payload === 'string'
        ? { name: payload, description: '' }
        : (payload || {});

    if(typeof name !== 'string' || name.trim() === ''){
        return {
            success: false,
            message: "Invalid category name"
        };
    }
    try{
        const data = await http (`/categories`, {
            method: "POST",
            body: JSON.stringify({ name, description })
        });
        return data;
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export const createCategories = createCategory;

export async function updateCategory(id, name, description = ''){
    if(typeof name !== 'string' || name.trim() === ''){
        return {
            success: false,
            message: "Invalid category name"
        };
    }
    try{
        const data = await http (`/categories/${id}`, {
            method: "PUT",
            body: JSON.stringify({name, description})
        });
        return data;
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}

export async function deleteCategory(id) {
    try{
        const data = await http (`/categories/${id}`, {
            method: "DELETE",
        });
        return data;
    } catch (err) {
        return {
            success: false,
            message: err.message
        };
    }
}
